File: src/test/java/helpers/HarCaptureTest.java
package helpers;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.logging.LogEntries;
import org.openqa.selenium.logging.LogEntry;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Tag("unit")
class HarCaptureTest {

    @Test
    void toHarBuildsEntriesFromPerformanceLogs() {
        LogEntries entries = fixtureEntries();

        String har = HarCapture.toHar(entries);
        assertTrue(har.contains("1.2"), () -> "HAR missing version: " + har);
        assertTrue(har.contains("example.com"), () -> "HAR missing url: " + har);
        assertTrue(har.contains("200"), () -> "HAR missing status: " + har);
        assertTrue(har.contains("1280"), () -> "HAR missing loadingFinished size: " + har);
        assertTrue(HarCapture.supportsBrowser("chrome"));
        assertTrue(!HarCapture.supportsBrowser("firefox"));
    }

    @Test
    void toHarDefaultMetaOmitsContentText() {
        LogEntries entries = fixtureEntries();

        String harDefault = HarCapture.toHar(entries);
        String harMeta = HarCapture.toHar(entries, HarCapture.HarContentMode.META);

        HarStats statsDefault = HarStats.fromBytes(
                "unit-meta-default",
                Path.of("unit-meta-default.har"),
                harDefault.getBytes(StandardCharsets.UTF_8));
        HarStats statsMeta = HarStats.fromBytes(
                "unit-meta",
                Path.of("unit-meta.har"),
                harMeta.getBytes(StandardCharsets.UTF_8));

        assertEquals(0, statsDefault.withContentText, "default toHar must stay meta (no content.text)");
        assertEquals(0, statsMeta.withContentText, "explicit META must omit content.text");
        assertFalse(harDefault.contains("\"text\""), () -> "default HAR leaked content.text: " + harDefault);
        assertFalse(harMeta.contains("\"text\""), () -> "META HAR leaked content.text: " + harMeta);
    }

    @Test
    void toHarBodiesIncludesSyntheticContentText() {
        LogEntries entries = fixtureEntries();
        // Performance logs have no bodies; BODIES mode consumes CDP (or synthetic) payloads by requestId.
        Map<String, HarCapture.CapturedBody> bodies = Map.of(
                "r1", new HarCapture.CapturedBody("<html>synthetic-body</html>", false));

        String har = HarCapture.toHar(entries, HarCapture.HarContentMode.BODIES, bodies);
        HarStats stats = HarStats.fromBytes(
                "unit-bodies",
                Path.of("unit-bodies.har"),
                har.getBytes(StandardCharsets.UTF_8));

        assertTrue(stats.withContentText > 0, "BODIES + synthetic fixture must yield content.text");
        assertTrue(har.contains("synthetic-body"), () -> "HAR missing body text: " + har);
        assertFalse(har.contains("\"encoding\""), "plain text body must not set encoding=base64");
    }

    @Test
    void toHarBodiesBase64SetsEncoding() {
        LogEntries entries = fixtureEntries();
        Map<String, HarCapture.CapturedBody> bodies = Map.of(
                "r1", new HarCapture.CapturedBody("aGVsbG8=", true));

        String har = HarCapture.toHar(entries, HarCapture.HarContentMode.BODIES, bodies);

        assertTrue(har.contains("aGVsbG8="), () -> "HAR missing base64 body: " + har);
        assertTrue(har.contains("base64"), () -> "HAR missing encoding=base64: " + har);
    }

    @Test
    void toHarBodiesWithoutPayloadStaysMetaLike() {
        LogEntries entries = fixtureEntries();

        String har = HarCapture.toHar(entries, HarCapture.HarContentMode.BODIES, Map.of());
        HarStats stats = HarStats.fromBytes(
                "unit-bodies-empty",
                Path.of("unit-bodies-empty.har"),
                har.getBytes(StandardCharsets.UTF_8));

        assertEquals(0, stats.withContentText, "BODIES without CDP/synthetic payload must stay meta-like");
    }

    @Test
    void finishedRequestIdsFromLoadingFinished() {
        assertEquals(List.of("r1"), HarCapture.finishedRequestIds(fixtureEntries()));
    }

    private static LogEntries fixtureEntries() {
        long now = System.currentTimeMillis();
        String requestMsg = """
                {"message":{"method":"Network.requestWillBeSent","params":{"requestId":"r1","timestamp":1.0,"wallTime":1700000000.0,"request":{"url":"https://example.com/","method":"GET","headers":{"Accept":"*/*"}}}}}
                """.trim();
        String responseMsg = """
                {"message":{"method":"Network.responseReceived","params":{"requestId":"r1","response":{"status":200,"statusText":"OK","mimeType":"text/html","headers":{"content-type":"text/html"},"protocol":"http/1.1","encodedDataLength":42}}}}
                """.trim();
        String finishedMsg = """
                {"message":{"method":"Network.loadingFinished","params":{"requestId":"r1","timestamp":1.05,"encodedDataLength":1280}}}
                """.trim();

        return new LogEntries(List.of(
                new LogEntry(Level.INFO, now, requestMsg),
                new LogEntry(Level.INFO, now + 1, responseMsg),
                new LogEntry(Level.INFO, now + 2, finishedMsg)
        ));
    }
}

File: internal/helpers/harcapture_test.go
package helpers_test

import (
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/qa-guru/selenoid-tests/internal/allurex"
	"github.com/qa-guru/selenoid-tests/internal/helpers"
)

func TestHarCapture_ToHarBuildsEntriesFromPerformanceLogs(t *testing.T) {
	allurex.Run(t, harMeta("toHar builds entries from performance logs"), func(a *allurex.A) {
		a.Step("toHar", func() {
			har := helpers.ToHar(fixtureEntries(), helpers.HarMeta, nil)
			require.Contains(t, har, "1.2")
			require.Contains(t, har, "example.com")
			require.Contains(t, har, "200")
			require.Contains(t, har, "1280")
			require.True(t, helpers.SupportsBrowser("chrome"))
			require.False(t, helpers.SupportsBrowser("firefox"))
		})
	})
}

func TestHarCapture_ToHarDefaultMetaOmitsContentText(t *testing.T) {
	allurex.Run(t, harMeta("toHar default/meta omits content.text"), func(a *allurex.A) {
		a.Step("compare modes", func() {
			entries := fixtureEntries()
			harDefault := helpers.ToHar(entries, helpers.HarMeta, nil)
			harMeta := helpers.ToHar(entries, helpers.HarMeta, nil)

			statsDefault := helpers.HarStatsFromBytes("unit-meta-default", []byte(harDefault))
			statsMeta := helpers.HarStatsFromBytes("unit-meta", []byte(harMeta))

			require.Equal(t, 0, statsDefault.WithContentText)
			require.Equal(t, 0, statsMeta.WithContentText)
			require.NotContains(t, harDefault, `"text"`)
			require.NotContains(t, harMeta, `"text"`)
		})
	})
}

func TestHarCapture_ToHarBodiesIncludesSyntheticContentText(t *testing.T) {
	allurex.Run(t, harMeta("toHar BODIES includes synthetic content.text"), func(a *allurex.A) {
		a.Step("bodies mode", func() {
			bodies := map[string]helpers.CapturedBody{
				"r1": {Text: "<html>synthetic-body</html>", Base64Encoded: false},
			}
			har := helpers.ToHar(fixtureEntries(), helpers.HarBodies, bodies)
			stats := helpers.HarStatsFromBytes("unit-bodies", []byte(har))
			require.Greater(t, stats.WithContentText, 0)
			require.Contains(t, har, "synthetic-body")
			require.NotContains(t, har, `"encoding"`)
		})
	})
}

func TestHarCapture_ToHarBodiesBase64SetsEncoding(t *testing.T) {
	allurex.Run(t, harMeta("toHar BODIES base64 sets encoding"), func(a *allurex.A) {
		a.Step("base64 body", func() {
			bodies := map[string]helpers.CapturedBody{
				"r1": {Text: "aGVsbG8=", Base64Encoded: true},
			}
			har := helpers.ToHar(fixtureEntries(), helpers.HarBodies, bodies)
			require.Contains(t, har, "aGVsbG8=")
			require.Contains(t, har, "base64")
		})
	})
}

func TestHarCapture_ToHarBodiesWithoutPayloadStaysMetaLike(t *testing.T) {
	allurex.Run(t, harMeta("toHar BODIES without payload stays meta-like"), func(a *allurex.A) {
		a.Step("empty bodies", func() {
			har := helpers.ToHar(fixtureEntries(), helpers.HarBodies, map[string]helpers.CapturedBody{})
			stats := helpers.HarStatsFromBytes("unit-bodies-empty", []byte(har))
			require.Equal(t, 0, stats.WithContentText)
		})
	})
}

func TestHarCapture_FinishedRequestIdsFromLoadingFinished(t *testing.T) {
	allurex.Run(t, harMeta("finishedRequestIds from loadingFinished"), func(a *allurex.A) {
		a.Step("ids", func() {
			require.Equal(t, []string{"r1"}, helpers.FinishedRequestIds(fixtureEntries()))
		})
	})
}

func harMeta(name string) allurex.Meta {
	return allurex.Meta{
		Name:      name,
		Package:   "helpers.HarCaptureTest",
		Layer:     "unit",
		Component: "selenoid",
		Suite:     "unit",
		Tags:      []string{"unit"},
	}
}

func fixtureEntries() []string {
	return []string{
		`{"message":{"method":"Network.requestWillBeSent","params":{"requestId":"r1","timestamp":1.0,"wallTime":1700000000.0,"request":{"url":"https://example.com/","method":"GET","headers":{"Accept":"*/*"}}}}}`,
		`{"message":{"method":"Network.responseReceived","params":{"requestId":"r1","response":{"status":200,"statusText":"OK","mimeType":"text/html","headers":{"content-type":"text/html"},"protocol":"http/1.1","encodedDataLength":42}}}}`,
		`{"message":{"method":"Network.loadingFinished","params":{"requestId":"r1","timestamp":1.05,"encodedDataLength":1280}}}`,
	}
}
